src/tsdb.c

```diff
@@ -224,75 +224,85 @@ int SeriesAddSample(Series *series, api_timestamp_t timestamp, double value) {
 }
 
-SeriesIterator SeriesQuery(Series *series, api_timestamp_t minTimestamp, api_timestamp_t maxTimestamp) {
-    SeriesIterator iter;
+int SeriesQuery(Series *series, SeriesIterator *iter, 
+                            api_timestamp_t minTimestamp, api_timestamp_t maxTimestamp) {
     timestamp_t rax_key;
-    iter.series = series;
+    iter->series = series;
+    ChunkFuncs *funcs = iter->series->funcs;
+    
     // get the rightmost chunk whose base timestamp is smaller or equal to minTimestamp
     seriesEncodeTimestamp(&rax_key, minTimestamp);
-    iter.dictIter = RedisModule_DictIteratorStartC(series->chunks, "<=", &rax_key, sizeof(rax_key));
-
-    // if no such chunk exists, we will start the search from the first chunk
-    if (!RedisModule_DictNextC(iter.dictIter, NULL, (void*)&iter.currentChunk))
-    {
-        RedisModule_DictIteratorReseekC(iter.dictIter, "^", NULL, 0);
-        RedisModule_DictNextC(iter.dictIter, NULL, (void*)&iter.currentChunk);
+    iter->dictIter = RedisModule_DictIteratorStartC(series->chunks, "<=", &rax_key, sizeof(rax_key));
+    RedisModule_DictNextC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+    
+    // iterate to the first relevant chunk
+    void *dictResult = (void *)TRUE;
+    while (dictResult) {
+        if (funcs->GetLastTimestamp(iter->currentChunk) < minTimestamp) {
+            dictResult = RedisModule_DictNextC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+        } else {
+            iter->chunkIterator = funcs->NewChunkIterator(iter->currentChunk);
+            iter->minTimestamp = minTimestamp;
+            iter->maxTimestamp = maxTimestamp;
+            return REDISMODULE_OK;
+        }
     }
-    iter.chunkIteratorInitialized = FALSE;
-    iter.chunkIterator = NULL;
-    iter.minTimestamp = minTimestamp;
-    iter.maxTimestamp = maxTimestamp;
-    return iter;
+    SeriesIteratorClose(iter);
+    return REDISMODULE_ERR;
 }
 
-void SeriesIteratorClose(SeriesIterator *iterator) {
-    iterator->series->funcs->FreeChunkIterator(iterator->chunkIterator);
-    RedisModule_DictIteratorStop(iterator->dictIter);
-}
+ChunkResult SeriesIteratorGetFirst(SeriesIterator *iterator, Sample *sample) {
+    ChunkFuncs *funcs = iterator->series->funcs;
+    ChunkResult res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
 
-int SeriesIteratorGetNext(SeriesIterator *iterator, Sample *currentSample) {
-    Sample internalSample;
-    while (iterator->currentChunk != NULL)
-    {
-        Chunk_t *currentChunk = iterator->currentChunk;
-        ChunkFuncs *funcs = iterator->series->funcs;
-        if (funcs->GetLastTimestamp(currentChunk) < iterator->minTimestamp)
-        {
-            if (!RedisModule_DictNextC(iterator->dictIter, NULL, (void*)&iterator->currentChunk)) {
-                iterator->currentChunk = NULL;
+    if (funcs->GetNumOfSample(iterator->currentChunk) == 0) {
+        return CR_END;
+    }
+
+    if (sample->timestamp < iterator->minTimestamp) {
```

Comment:
@ashtul there's a potential bug here, you depend on `sample` having a correct timestamp that I would guess point to the previous one, but it really depends on the context of the caller.
For example on the first time you use the iterator and call this the sample would be empty (0) or might have some garbage it can cause un predicated behaviour.

Reply:
@danni-m This is a fresh next sample that was retrieved line 256.
If a sample return CR_OK, it is valid.

We can switch the `while` with `if` maybe to have a better checking for ChunkResult

Reply:
Changed the code to check the sample is valid before using its values.